--- a/dwl.c
+++ b/dwl.c
@@ -256,6 +256,7 @@ typedef struct {
 
 /* function declarations */
 static void applybounds(Client *c, struct wlr_box *bbox);
+static void col(Monitor *m);
 static void applyrules(Client *c);
 static void applyexclusive(struct wlr_box *usable_area,
 		uint32_t anchor, int32_t exclusive,
@@ -311,6 +312,7 @@ static void gpureset(struct wl_listener *listener, void *data);
 static void handlesig(int signo);
 static void incnmaster(const Arg *arg);
 static void inputdevice(struct wl_listener *listener, void *data);
+static void grid(Monitor *m);
 static int keybinding(uint32_t mods, xkb_keysym_t sym);
 static void keypress(struct wl_listener *listener, void *data);
 static void keypressmod(struct wl_listener *listener, void *data);
@@ -991,6 +993,69 @@ chvt(const Arg *arg)
 		wlr_session_change_vt(session, arg->ui);
 }
 
+static void
+col(Monitor *m)
+{
+	Client *c;
+	unsigned int i, n = 0, x = 0, w = 0;
+
+	wl_list_for_each(c, &clients, link)
+		if (VISIBLEON(c, m) && !c->isfloating && !c->isfullscreen)
+			n++;
+	if (n == 0)
+		return;
+
+	i = 0;
+	wl_list_for_each(c, &clients, link) {
+		if (!VISIBLEON(c, m) || c->isfloating || c->isfullscreen)
+			continue;
+
+		w = (m->w.width - x) / (n - i);
+		resize(c, (struct wlr_box){.x = m->w.x + x, .y = m->w.y,
+			.width = w, .height = m->w.height}, 0);
+		x += w;
+		i++;
+	}
+}
+
+static void
+grid(Monitor *m)
+{
+	Client *c;
+	unsigned int i, n = 0, cx, cy, cw, ch, aw, ah, cols, rows;
+
+	wl_list_for_each(c, &clients, link)
+		if (VISIBLEON(c, m) && !c->isfloating && !c->isfullscreen)
+			n++;
+	if (n == 0)
+		return;
+
+	for (rows = 0; rows <= n / 2; rows++)
+		if (rows * rows >= n)
+			break;
+	if (rows == 0)
+		rows = 1;
+	cols = (rows && (rows - 1) * rows >= n) ? rows - 1 : rows;
+	if (cols == 0)
+		cols = 1;
+
+	ch = m->w.height / rows;
+	cw = m->w.width / cols;
+	i = 0;
+	wl_list_for_each(c, &clients, link) {
+		if (!VISIBLEON(c, m) || c->isfloating || c->isfullscreen)
+			continue;
+
+		cx = m->w.x + (i / rows) * cw;
+		cy = m->w.y + (i % rows) * ch;
+		aw = (i >= rows * (cols - 1)) ? m->w.width - cw * cols : 0;
+		ah = ((i + 1) % rows == 0) ? m->w.height - ch * rows : 0;
+		resize(c, (struct wlr_box){.x = cx, .y = cy,
+			.width = cw + aw, .height = ch + ah}, 0);
+		i++;
+	}
+}
+
 void
 cleanup(void)
 {
